Feature/add trello end point

Danielseddian opened the pull request:
Add endpoint and base function for trello webhook

Danielseddian committed:
Add trello endpoint and base function for webhook

Danielseddian committed:
Add trello endpoint and base function for webhook

Danielseddian committed:
Add trello endpoint and base function for webhook

veselij commented on src/run_webhook_api.py:
```diff
@@ -38,9 +38,25 @@ async def telegram(request: Request) -> Response:
+        return Response()
+    elif request.method == "POST":
+        if request.headers and await request.json():
+            message = "Информация с сайта Trello успешно получена и обработана."
+            await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)
```
закоментируй пожалуйста эти строки, так как у нас в конфете CHAT_ID, напиши просто в качестве коммента пример отправки сообщения и эти две закоменченные строчки

veselij reviewed (changes requested):
в целом ок, только пофикси замечания по пеп, чтобы можно было мержить.
и положи пример json файла, того что приходит от трелло и в readme добавь описание, что вот так запускаем бота, берем json пример и curlом отправляем его на слушающий энд Поинт. чтобы тот кто будет делать след задание по обработке этого json понимал как его получить и в каком он виде

Danielseddian committed:
Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

veselij commented on README.md:
```diff
@@ -7,8 +7,77 @@
+    "closed": false,
+    "idOrganization": "4e1452614e4b8698470000e0",
+    "pinned": true,
+    "url": "https://trello.com/b/nC8QJJoZ/trello-development",
```
положи пожалуйста просто файл .json в папочку src/example, а тут ссылку на него

Danielseddian replied:
Ага, понял, поправил

veselij reviewed (approved):
поправь readme пожалуйста

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

AntonZelinsky commented on src/run_webhook_api.py:
```diff
@@ -38,9 +38,26 @@ async def telegram(request: Request) -> Response:
+    :param request: Trello request
+    :return: Response "ok"
+    """
+    if request.method == "HEAD":
```
А для чего метод хэд используется? При инициализации трелло?Добавь кометарий об этом, можно даже в  докстринг.

AntonZelinsky commented on src/run_webhook_api.py:
```diff
@@ -38,9 +38,26 @@ async def telegram(request: Request) -> Response:
 routes = [
     Route("/telegram", telegram, methods=["POST"]),
     Route("/health", health, methods=["GET"]),
+    Route("/trelloCallback", trello_call_back, methods=["POST", "HEAD"]),
```
Давай немного переименуем методы и добавим это в документик с соглашением
```suggestion
    Route("/telegramWebhookApi", telegram_webhook_api, methods=["POST"]),
    Route("/healthcheck", healthcheck_api, methods=["GET"]),
    Route("/trelloWebhookApi", trello_webhook_api, methods=["POST", "HEAD"]),
```
что для всех апи нужно добавлять окончание _апи

AntonZelinsky commented on src/run_webhook_api.py:
```diff
@@ -38,9 +38,26 @@ async def telegram(request: Request) -> Response:
+        #     message = "Информация с сайта Trello успешно получена и обработана."
+        #     await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)
+        return Response("Message received")
+    return Response("Something wrong", status_code=405)
```
Судя по ограничениям роутов, этот кейс никогда не будет выполнен.

AntonZelinsky commented on README.md:
```diff
@@ -7,8 +7,19 @@
+payload = {'query': json.dumps({"test_key": "test_value"})}
+url = "https://www.test.com/test_endpoint"
+
+response = requests.post(url, data=payload)
```
Что этот пример иллюстрирует и зачем он тут я не понял..

veselij replied:
Денис добавь прям пример как Curl отправить example json (тот что у нас в папке примеров лежит) на наш endpoint trello, чтобы просто копируешь отсюда команду в терминал и получаешь запрос на endoint

AntonZelinsky commented on src/run_webhook_api.py:
```diff
@@ -38,9 +38,26 @@ async def telegram(request: Request) -> Response:
+    elif request.method == "POST":
+        # if request.headers and await request.json():
+        #     message = "Информация с сайта Trello успешно получена и обработана."
+        #     await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)
```
Сделай тогда уже получение из запроса каких-то данных, даже самой простой id, но это уже иллюстрацию работы. Просто я подозреваю, что `await request.json()` возвращает какие-то данные и они попадают в иф и всё, выходит, этот пример немного некорректный.
Добавь логер и логируй что-нибудь из запроса туда.

Следующим заданием для тебя будет написать дата классы, хотя бы 1-2 штуки из иерархии и десириализовать в них json, но только нужно сделать это нужно уже в другом ПРе, чтобы эту задачку быстрее двинуть вперёд.
Сторонние библиотеки хотелось бы не подключать если есть более-менее вменяемые способы сделать это без них

AntonZelinsky reviewed (commented):
[Суда по всему](https://developer.atlassian.com/cloud/trello/guides/rest-api/webhooks/), можно из нашего кода сообщать трелло, что мы подключаем вебхук и нужно на него слать запросы. И именно это нужно сделать, так как руками каждый раз это делать никто не будет.

veselij commented:
> [Суда по всему](https://developer.atlassian.com/cloud/trello/guides/rest-api/webhooks/), можно из нашего кода сообщать трелло, что мы подключаем вебхук и нужно на него слать запросы. И именно это нужно сделать, так как руками каждый раз это делать никто не будет.

но это по идее отдельная задача? это же не в эндпоинт вебхука? там еще надо для этого запроса сделать переменные окружения с токенами, и делать этот запрос убедившись, что наш сервер поднят, так как трелло после получения запроса на активацию, проверяет HEAD запросом переданный ей URL, "рабочий" ли он

AntonZelinsky commented:
Окей, пусть будет отдельной задачей. Но это всё в рамках одной общей задачи делается, пусть и из двух подзадач. Так как одно не имеет смысла без другого.
@Danielseddian, в общем, мелочи поправь в этом ПРе, а то что по крупянее в следующую задачу выноси.

Danielseddian commented:
Очень интересно, завтра все гляну, сегодня не дома

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

Danielseddian committed:
Дополнил README.md, добавил информацию по request'у от trello и curl на другой endpoint

Проверка headers и body
if request.headers and await request.json():
    message = "Информация с сайта Trello успешно получена и обработана."
    await request.app.state.bot_app.bot.send_message(chat_id=config.CHAT_ID, text=message)

veselij commented on src/run_webhook_api.py:
```diff
@@ -24,23 +25,37 @@ async def stop_bot() -> None:
-async def health(request: Request) -> PlainTextResponse:
+async def healthcheck_api(request: Request) -> PlainTextResponse:
     message = f"Бот запущен и работает. Сообщение получено по запросу на Api сервера {config.WEBHOOK_URL}"
     chat_id = config.CHAT_ID
```
chat_id больше нет у нас, упадет же при запуске, сделай просто логирование в консоль

veselij commented on src/run_webhook_api.py:
```diff
@@ -24,23 +25,37 @@ async def stop_bot() -> None:
+    :return: Response "ok"
+    """
+    if request.method == "POST":
+        trello_model_id: int = dict(await request.json()).get("model", {}).get("id", 0)
```
как-то очень сложно в одну строку, получи сначала респонс, а потом просто выбери через [model][id]

veselij commented on src/run_webhook_api.py:
```diff
@@ -24,23 +25,37 @@ async def stop_bot() -> None:
+    """
+    if request.method == "POST":
+        trello_model_id: int = dict(await request.json()).get("model", {}).get("id", 0)
+        logging.info(f"Got trello request, model id: {trello_model_id}" if trello_model_id else "Got empty request")
```
и просто залогируй без if else. это же для примера, не надо тут логику накручивать. это все десериализации на датакалассы будет

veselij commented on src/run_webhook_api.py:
```diff
@@ -24,23 +25,37 @@ async def stop_bot() -> None:
+    if request.method == "POST":
+        trello_model_id: int = dict(await request.json()).get("model", {}).get("id", 0)
+        logging.info(f"Got trello request, model id: {trello_model_id}" if trello_model_id else "Got empty request")
+        return Response("Message received", status_code=200)
+    return Response("ok", status_code=200)
```
статус коды лучше писать через httpx.codes.OK - так как понятно что за код. с 200 то лего, а с каким то 401 надо вспоминать, что это. лучше когда явно и однообразно написано

veselij commented on src/core/logger.py:
```diff
@@ -0,0 +1,7 @@
+import logging
+
+logging.basicConfig(
+    filename="bot.log",
+    level=logging.INFO,
+    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
+)
```
логирование просто в консоль надо, файл не надо создавать

veselij reviewed (changes requested):
поправь еще пожалуйста пару замечаний, и все таки надо запускать свой код перед отправкой PR

Danielseddian committed:
Updated healthcheck_api and trello_webhook_api.

Danielseddian committed:
Updated healthcheck_api and trello_webhook_api.
